improve examples/read.c and examples/blink.c

vitalogy opened the pull request:
hi,

i have made some changes to read.c. 
Both GPIO's must be set manual by the user, there are no default GPIO's and both must be numeric. Here are some examples:

```
# wiringx-read
Usage: read GPIO GPIO
   first GPIO to write = output
   second GPIO to read from = input
Example: read 4 23

# wiringx-read 24
Usage: read GPIO GPIO
   first GPIO to write = output
   second GPIO to read from = input
Example: read 4 23

# wiringx-read 24h 25
read: Invalid GPIO 24h

# wiringx-read 24 25j
read: Invalid GPIO 25j

# wiringx-read 24 25 7
Usage: read GPIO GPIO
    first GPIO to write = output
   second GPIO to read from = input
Example: read 4 23

# wiringx-read 24 25
running on a raspberrypi
Writing to GPIO 24: High
Reading from GPIO 25: 1
Writing to GPIO 24: Low
Reading from GPIO 25: 0
Writing to GPIO 24: High
Reading from GPIO 25: 1
Writing to GPIO 24: Low
Reading from GPIO 25: 0
Writing to GPIO 24: High
Reading from GPIO 25: 1
Writing to GPIO 24: Low
Reading from GPIO 25: 0
Writing to GPIO 24: High
Reading from GPIO 25: 0
```

So make the same changes to blink.c, there is no more a default GPIO, if the user does not set one.

```
# wiringx-blink 
Usage: blink GPIO
   GPIO is the GPIO to write
Example: blink 10

# wiringx-blink Three
blink: Invalid GPIO Three

# wiringx-blink 24h
blink: Invalid GPIO 24h

# wiringx-blink 24 25
Usage: blink GPIO
   GPIO is the GPIO to write
Example: blink 10

# wiringx-blink 24
running on a raspberrypi
```

As i reading through my own words, i see that the name of the program in the output should be replaced by argv[0], so name will be the same as you calling the program.

Please take a look and say what you are thinking.

Commit:
examples: GPIO's must manual set by the user + error handling

CurlyMoo commented on examples/read.c:
```diff
@@ -4,19 +4,61 @@
 
-int main(void) {
+char *usage = "Usage: read GPIO GPIO\n"
+				  "       first GPIO to write = output\n"
```
I would make this:
`first GPIO to write to`

vitalogy commented:
Made the change you suggested and set additionally argv[0] for the calling program.

Is the way i did the rebase okay?

CurlyMoo commented on examples/blink.c:
```diff
@@ -4,32 +4,46 @@
 
-int main(int argc, char *argv[]) {
-	int pin = 0;
+char *usage = "Usage: blink GPIO\n"
```
You missed a `argv[0]`

CurlyMoo commented on examples/read.c:
```diff
@@ -4,19 +4,61 @@
 #include "wiringX.h"
 
-int main(void) {
+char *usage = "Usage: read GPIO GPIO\n"
```
You missed a `argv[0]`

CurlyMoo commented on examples/read.c:
```diff
@@ -4,19 +4,61 @@
+char *usage = "Usage: read GPIO GPIO\n"
+			"       first GPIO to write to = output\n"
+			"       second GPIO to read from = input\n"
+			"Example: read 4 23";
```
You missed a `argv[0]`

CurlyMoo commented on examples/blink.c:
```diff
@@ -4,32 +4,46 @@
-	int pin = 0;
+char *usage = "Usage: blink GPIO\n"
+			"       GPIO is the GPIO to write\n"
+			"Example: blink 10";
```
You missed a `argv[0]`

vitalogy commented:
Yes, I have noticed this ... but can i set argv[0] in this character string? Will this then be replaced?

CurlyMoo commented:
No, you need to sprintf it.

vitalogy commented:
Okay ... is something like this `sprintf(str, usage, argv[0], argv[0]);`
Will have to take a look at sprintf.

Commit:
examples: return the name of called program

vitalogy commented:
Let me know, if it's not okay. And thanks for your patience.

CurlyMoo commented on examples/read.c:
```diff
@@ -4,19 +4,62 @@
 #include "wiringX.h"
 
-int main(void) {
+char *usage = "Usage: %s GPIO GPIO\n"
```
Type dual GPIO

vitalogy commented:
You have to give two GPIO's... the first for output and the second for the input

Commit:
examples/read.c: forget to set the newline

vitalogy commented:
```
 $ ./read
Usage: ./read GPIO GPIO
   first GPIO to write to = output
   second GPIO to read from = input
Example: ./read 4 23
```

Commit:
examples/blink.c: forget to set the newline

CurlyMoo commented:
Aha. You're right.

CurlyMoo commented on examples/blink.c:
```diff
@@ -4,32 +4,47 @@
-int main(int argc, char *argv[]) {
-	int pin = 0;
+char *usage = "Usage: %s GPIO\n"
+			"       GPIO is the GPIO to write\n"
```
`... to write to\n`

CurlyMoo replied:
Also, make sure your indentations are correct.

CurlyMoo commented on examples/blink.c:
```diff
@@ -4,32 +4,47 @@
+	// found at http://stackoverflow.com/questions/17292545
+	for(i=1; i<argc; i++) {
+		n = sscanf(argv[i], "%d%c", &value, &ch);
+		if (n != 1) {
```
This is not wiringX programming style `if(...) {`

CurlyMoo commented on examples/blink.c:
```diff
@@ -4,32 +4,47 @@
+	// expect 1 argument for the programm blink ('blink gpio')
+	// these are argv[0-1], so argc = 2
+	if(argc != 2) {
+		sprintf(str, usage, argv[0], argv[0]);
```
Even better: `snprintf(str, 99, usage, argv[0], argv[0]);`

CurlyMoo commented on examples/blink.c:
```diff
@@ -4,32 +4,47 @@
-	wiringXSetup();
+int main(int argc, char *argv[]) {
+	int gpio;
+	int i, n, value;
```
Always initialize your variables.

CurlyMoo commented on examples/blink.c:
```diff
@@ -4,32 +4,47 @@
+int main(int argc, char *argv[]) {
+	int gpio;
+	int i, n, value;
+	char ch, str [100];
```
`ch = 0, str[100];`
`memset(str, '\0', 100);`

vitalogy replied:
will do

CurlyMoo commented on examples/read.c:
```diff
@@ -4,19 +4,62 @@
+	// first, check for valid, numeric arguments
+	// found at http://stackoverflow.com/questions/17292545
+	for(i=1; i<argc; i++) {
+		n = sscanf(argv[i], "%d%c", &value, &ch);
```
http://www.tutorialspoint.com/c_standard_library/c_function_isdigit.htm

vitalogy replied:
I have played with isdigt but I found is to complicated, especially for me :-(

I have this code now:

```
int i, val;
char *endptr, *str;

for(i=1; i<argc; i++) {
   str = argv[i];
   val = (int)strtol(str, &endptr, 10);
   if((*endptr != '\0') || (val < 0 || val >= 128)) {
      printf("%s: Invalid GPIO %s\n", argv[0], str);
      return -1;
   }
}
```

Also with these, the following segmentation fault is gone with a GPIO above 127.

```
./read 128 6
running on a raspberrypi
Segmentation fault (core dumped)

./read 127 6
running on a raspberrypi
./read: Invalid GPIO 127 for output
```

now a GPIO above 127 and GPIO with mistypes, gives the following:

```
./read 544 22 
./read: Invalid GPIO 544

./read 11g 22
./read: Invalid GPIO 11g
```

Is this a better solution?

CurlyMoo replied:
Not sure what so difficult about `isdigit`:

```
#include <ctype.h>
#include <stdio.h>

int main(void) {
        char *str = "1234a";
        int invalid = 0;
        while(*str != '\0') {
                if(!isdigit(*str)) {
                        invalid = 1;
                }
                str++;
        }
        if(invalid == 1) {
                printf("not a digit\n");
        }
        return 0;
}
```

If you want to avoid `isdigit` you can also do this:

```
#include <stdio.h>

int main(void) {
        char *str = "1234a";
        int invalid = 0;
        while(*str != '\0') {
                if(*str < 48 || *str > 57) {
                        invalid = 1;
                }
                str++;
        }
        if(invalid == 1) {
                printf("not a digit\n");
        }
        return 0;
}
```

Commit:
examples: changes to read.c and blink.c

vitalogy commented:
next try :-)
- changed the things as you suggested
- added status message to blink

The problem with the tabsize is browser or github related (I don't know), but you can append `?ts=4` to the URL. I have put the usage message in a newline and indent, so let me know, if this is okay!
